### PR TITLE
Unify workspace/global AppHost describe --follow streaming

### LINKED ISSUE #18123: Unify workspace/global AppHost `describe --follow` streaming in VS Code extension
`AppHostDataRepository` currently has separate `describe` pipelines for workspace mode (`_startDescribeWatch`) and global mode (`_startGlobalDescribe` fan-out), even though both execute the same CLI shape and very similar stream lifecycle logic. This duplication increases maintenance overhead and makes multi-AppHost workspace behavior harder to reason about.

### Goal
Refactor to a single describe-stream manager keyed by `appHostPath`, with workspace/global behavior expressed as policy over one shared stream set.

### Desired behavior after refactor
Use the workspace-filtered running set as input to unified describe reconciliation:
- Workspace desired describe paths:
  - paths from the current workspace running set (root-filtered while discovery is pending; then effectively `psRunning ∩ workspaceCandidates` after candidates are known).
- Global desired describe paths:
  - all running `ps` AppHost paths.
- Selected workspace AppHost controls projection/UI (for example, `workspaceResources` source), not stream eligibility.

### Scope
- Consolidate duplicated describe process lifecycle logic:
  - spawn args construction
  - `--include-disabled-commands` capability gating and fallback
  - NDJSON parsing + resource map updates
  - restart/backoff, teardown, and cleanup
- Keep workspace candidate discovery as-is (`aspire ls` + existing fallbacks + config augmentation).
- Preserve existing tree rendering semantics and selected-host behavior.

### Non-goals
- Changing AppHost discovery source of truth.
- Changing workspace/global UX beyond parity with current behavior.

### Acceptance criteria
1. Workspace and global modes share one internal describe stream manager.
2. Workspace desired describe paths are derived from workspace-filtered running AppHosts:
   - pending discovery: workspace-root-filtered running paths
   - after discovery: running paths matching workspace candidates.
3. No regressions in:
   - selected workspace host retargeting
   - non-selected running workspace hosts receiving resources
   - global per-AppHost resource population
4. Existing tests pass; add/update tests to validate unified reconcile behavior in both modes.

### Suggested implementation notes
- Introduce a mode-aware “desired describe paths” reconcile step.
- Maintain one map of active describe streams keyed by `appHostPath`.
- Add a small projection layer for selected workspace host -> `workspaceResources`.
- Preserve workspace-specific no-data/compatibility error surfacing behavior.

### LINKED ISSUE #18579: Single-file AppHost describe ignores the open file and resolves from `process.cwd()`
When an AppHost file is opened with no workspace folder, the extension runs a path-less `aspire describe --follow` (the "single-file" describe). Because `spawnCliProcess` falls back to `vscode.workspace.workspaceFolders?.[0]?.uri.fsPath ?? process.cwd()` and the single-file describe sets no `workingDirectory` and no `--apphost`, the CLI resolves the AppHost from the extension host's `process.cwd()` — **not** from the AppHost file in the editor. Result: `describe --follow exited (code 0) without producing data; not auto-restarting`, even though a valid AppHost file is open.

**Expected:** The single-file describe should target the open AppHost file.

**Notes / suggested fix:** `AppHostFilePresenceWatcher` already identifies the AppHost editor via `getParserForDocument(editor.document)` but reports only a boolean to `setAppHostFileOpen`. Thread the document's `fsPath` through so the single-file describe can pass `--apphost <path>` (or use `path.dirname(path)` as `workingDirectory`) instead of relying on `process.cwd()`.

**Repro:** Open just `TestShop.AppHost/AppHost.cs` (no folder). Run the apphost. Observe the Output log: path-less `describe --follow` exits without data.

### PR DESCRIPTION
**Pull request by ellahathaway**
Fixes #18123
Fixes #18579

## Summary

Refactors VS Code extension AppHost runtime tracking so `aspire ps` is the single source of running AppHosts and every rendered running AppHost gets its own `aspire describe --follow --apphost <path>` stream.

This removes the split between workspace describe, global describe fan-out, and visible-editor-only AppHost tracking. Open AppHost tabs are now tracked by path, so single-file or out-of-workspace AppHosts can stay visible/followed while their tab is open.

This also lets running AppHosts and their resources render before workspace AppHost discovery finishes. The panel no longer has to wait for discovery before `ps`/`describe` data can populate running resources.

## Details

- Track open AppHost tabs by path instead of a single visible/open boolean.
- Keep `aspire ps` active when the AppHost panel is visible or at least one AppHost tab is open.
- Start describe streams only after `aspire ps` reports a matching running AppHost.
- Use one unified `_describeStreams` map for workspace, global, and open-tab AppHosts.
- Derive `workspaceAppHost` from `_workspaceAppHostPath` and the current `ps` state instead of storing separate workspace AppHost state.
- Allow running AppHosts and resources to appear before workspace discovery completes.
- Reconcile describe streams when view mode, open tabs, discovery, or `ps` output changes.
- Rename `aspire.globalAppHostsPollingInterval` to `aspire.appHostsPollingInterval`, with the old setting retained as a deprecated fallback.
- Clear workspace loading state when discovery completes even if `ps --follow` does not emit another event.
- Avoid stale resource data when describe streams stop.
- Suppress expected timeout warnings when intentionally stopping `describe --follow` streams.

Before:
https://github.com/user-attachments/assets/93c76a84-ee37-4698-9956-b6d6917d7c2d

After:
https://github.com/user-attachments/assets/c8c71087-5c07-4a1c-9eaa-e0ab828cf0b3

### CODE REVIEW

File: extension/src/views/AppHostDataRepository.ts [hunk 1]
```diff
@@ -135,6 +135,12 @@ interface GlobalDescribeStream {
     restartTimer: ReturnType<typeof setTimeout> | undefined;
     restartDelay: number;
     version: number;
+    receivedData: boolean;
+    startPending: boolean;
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/src/views/AppHostDataRepository.ts]: Fixed in https://github.com/microsoft/aspire/pull/18527/commits/48bba8821aec4a39a5e01c8dcc040755669d112b

File: extension/src/views/AppHostDataRepository.ts [hunk 2]
```diff
@@ -1032,40 +925,144 @@ export class AppHostDataRepository {
...
+        }
+
+        const desired = new Set(this._runningWorkspaceAppHosts.map(a => a.appHostPath));
+        const target = this._resolveWorkspaceDescribeTarget();
+        if (target !== undefined && !this._isRunningWorkspaceHost(target)) {
+            const existing = this._findDescribeStream(target);
+            if (!(existing && existing.receivedData)) {
+                desired.add(target);
             }
         }
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/src/views/AppHostDataRepository.ts]: Fixed in https://github.com/microsoft/aspire/pull/18527/commits/fbb4b05008206acd23d257697856e362068e9776

File: extension/src/views/AppHostDataRepository.ts [hunk 3]
```diff
@@ -1032,40 +925,144 @@ export class AppHostDataRepository {
...
+            }
+
+            // A stream that parked while its host was idle must restart now that the host is running.
+            // If it still produces no data it re-parks as `parkedWhileRunning: true`, so this fires
+            // exactly once on the idle->running edge instead of looping every poll.
+            const stream = this._describeStreams.get(key)!;
+            if (stream.parkedWhileRunning === false && this._isRunningWorkspaceHost(path)) {
+                this._stopDescribe(key);
+                this._startDescribe(path);
+            }
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/src/views/AppHostDataRepository.ts]: Fixed in https://github.com/microsoft/aspire/pull/18527/commits/832f99d7f070e35fd7ce897eae8f92b06dbdc085

**Comment by ellahathaway** [comment]
Needs https://github.com/microsoft/aspire/issues/18579

File: extension/package.json [hunk 4]
```diff
@@ -899,11 +899,19 @@
           "description": "%configuration.aspire.registerMcpServerInWorkspace%",
           "scope": "window"
         },
+        "aspire.appHostsPollingInterval": {
+          "type": "number",
+          "default": 30000,
+          "minimum": 1000,
+          "description": "%configuration.aspire.appHostsPollingInterval%",
+          "scope": "window"
+        },
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/package.json:908]: Fixed in 2bcda5d0d — added 'aspire.appHostsPollingInterval' to xpectedConfigurationKeys in packageSurface.e2e.test.ts (sorted position, after spire.appHostDiscoveryTimeoutMs). Verified compile-e2e passes.

[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/package.json:908]: Fixed in 2bcda5d0d — 'aspire.appHostsPollingInterval' is now included in `expectedConfigurationKeys`, so the deepStrictEqual package-surface audit matches the contributed configuration keys.

File: extension/package.nls.json [hunk 5]
```diff
@@ -43,7 +43,9 @@
   "configuration.aspire.closeDashboardOnDebugEnd": "Close the dashboard browser when the debug session ends. Works with debug browser options (Chrome, Edge, Firefox).",
   "configuration.aspire.enableDebugConfigEnvironmentLogging": "Include environment variables when logging debug session configurations. This can help diagnose environment-related issues but may expose sensitive information in logs.",
   "configuration.aspire.registerMcpServerInWorkspace": "Whether to register the Aspire MCP server when a workspace is open.",
-  "configuration.aspire.globalAppHostsPollingInterval": "Polling interval in milliseconds for fetching all running AppHosts (used in global view). Minimum: 1000.",
+  "configuration.aspire.appHostsPollingInterval": "Polling interval in milliseconds for fetching running AppHosts. Minimum: 1000.",
+  "configuration.aspire.globalAppHostsPollingInterval": "Polling interval in milliseconds for fetching running AppHosts. Minimum: 1000.",
+  "configuration.aspire.globalAppHostsPollingInterval.deprecated": "This setting has been renamed to aspire.appHostsPollingInterval.",
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/package.nls.json:48]: Already resolved in eb08a6e80 — the xlf was regenerated. `extension/loc/xlf/aspire-vscode.xlf` now has the updated configuration.aspire.globalAppHostsPollingInterval source text plus the two new trans-units (configuration.aspire.appHostsPollingInterval and configuration.aspire.globalAppHostsPollingInterval.deprecated), matching package.nls.json. This comment predates that regeneration.

File: extension/src/views/AppHostDataRepository.ts [hunk 6]
```diff
@@ -931,167 +906,139 @@ export class AppHostDataRepository {
...
+                        this._handlePsSnapshot(this._appHosts.filter(appHost => !isMatchingAppHostPath(appHost.appHostPath, appHostPath)));
+                        return;
+                    }
+
+                    // A stream that never produced resources and exits (cleanly or with an error) means the CLI
+                    // cannot describe the host. Surface it once through the describe banner
+                    const noDataError = this._getDescribeNoDataError(code, stream.nonJsonLines, stream.stderr);
+                    if (noDataError.message) {
+                        extensionLogOutputChannel.warn(`aspire describe --follow (--apphost ${appHostPath}) exited (code ${code}) without producing data.`);
+                        this._setDescribeError(noDataError.message, { compatibility: noDataError.isCompatibilityError });
```
Comment (joperezr) [inline extension/src/views/AppHostDataRepository.ts]: What happens if a non-selected apphost from the workspace exists here with an error? Will it still write to the shared banner?

Essentially the scenario I'm thinking about is a workspace with two Apphosts, and the selected one is A, if B has an issue in the stream call would there be a banner saying there is an issue and have that seem like it's a problem with A instead of B?

Mainly mentioning it since it seems like the old code was specifically not touching the banner in the non-selected case specifically for this reason, so it "don't pollute the global banner".

**Review by joperezr — APPROVED** [review]
Left only one minor comment but the rest looks really good! Thanks for re-architecting this, it looks much cleaner now. 😃

File: extension/src/views/AppHostDataRepository.ts (hunk 6 above)
Comment (ellahathaway) [reply extension/src/views/AppHostDataRepository.ts]: Thanks! Addressed in https://github.com/microsoft/aspire/pull/18527/commits/a3f180ab12aff5b5d8410fef426029009b3c3c50

File: extension/src/views/AppHostDataRepository.ts [hunk 7]
```diff
@@ -1858,108 +1701,76 @@ export class AppHostDataRepository {
...
             }
         }
+    }
 
-        this._attachGlobalResourcesToAppHosts();
+    private _isWorkspaceAppHost(appHost: AppHostDisplayInfo): boolean {
+        const isOpenAppHostPath = this._openAppHostPaths.some(openPath => isMatchingAppHostPath(appHost.appHostPath, openPath));
+        const isSelectedWorkspaceAppHostPath = isMatchingAppHostPath(appHost.appHostPath, this._workspaceAppHostPath);
+        const isWorkspaceCandidatePath = this._workspaceAppHostCandidatePaths.some(candidatePath => isMatchingAppHostPath(appHost.appHostPath, candidatePath));
+        return isOpenAppHostPath || isSelectedWorkspaceAppHostPath || isWorkspaceCandidatePath || isPathInWorkspace(appHost.appHostPath);
```
[the remark this answers is not in the document]
Comment (ellahathaway) [reply extension/src/views/AppHostDataRepository.ts:1792]: The membership set only ever contains hosts that aspire ps reports as currently running, so any host that reaches the isPathInWorkspace(...) clause is, by construction, a live process, which means it is buildable and built. The "non-buildable / discovery-excluded hosts get rendered and described" scenario can't occur through this path.

Given that, psRunning ∩ workspaceCandidates is actually the narrower, lossy option: aspire ls (static discovery) and aspire ps (live processes) can legitimately disagree — a host can be running under the workspace root without ls having enumerated it as a candidate (timing, a source-file host vs. resolved .csproj, a project ls didn't classify as buildable). Intersecting with candidates would hide a genuinely-running in-workspace AppHost from the workspace view, which is the real regression. That behavior is intentional and locked in by the existing tests (runtime state refresh uses ps fallback when active workspace has no discovered candidate and ...replays ps fallback after empty discovery) — they assert a running in-workspace host stays visible after an empty discovery, so they encode the intended contract rather than needing correction.

#18123's psRunning ∩ candidates criterion presumes candidates are a superset of the running in-workspace hosts, which isn't guaranteed. The projection here is deliberately psRunning filtered to (open-tab ∪ selected ∪ candidate ∪ under-workspace-root): every rendered host is real, and streaming describe for a genuinely-running in-workspace host is exactly the unified per-AppHost behavior we want. Keeping the unconditional root clause.

File: extension/src/editor/AppHostFilePresenceWatcher.ts [hunk 8]
```diff
@@ -59,27 +86,75 @@ export class AppHostFilePresenceWatcher implements vscode.Disposable {
...
+
+    private async _isOpenAppHostDocument(document: vscode.TextDocument): Promise<boolean> {
+        if (await getParserForDocument(document)) {
+            return true;
+        }
+
+        // A project AppHost (.csproj/.fsproj/.vbproj) has no inline resources to parse, but opening it
+        // should still surface and follow its running instance.
+        if (AppHostFilePresenceWatcher._projectAppHostExtensions.has(getFileExtension(document.uri.fsPath))) {
+            return projectContentsReferencesRunnableAspireAppHost(document.getText());
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/AppHostFilePresenceWatcher.ts:131]: Added focused coverage in f07a54428a for a backgrounded AppHost .csproj, exclusion of a regular SDK project, and project-file edits transitioning into and out of AppHost status after the debounce. I also verified the open AppHost and regular project cases in the desktop UI.

File: extension/src/views/AppHostDataRepository.ts [hunk 9]
```diff
@@ -1119,214 +1109,76 @@ export class AppHostDataRepository {
...
         }
-        this._attachGlobalResourcesToAppHosts();
-    }
 
-    private _attachGlobalResourcesToAppHosts(): void {
-        for (const appHost of this._appHosts) {
-            const stream = this._globalDescribeStreams.get(appHost.appHostPath);
-            appHost.resources = stream ? Array.from(stream.resources.values()) : null;
+        if (desired.size === 0 && this._describeErrorIsCompatibility) {
+            this._setDescribeError(undefined);
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/views/AppHostDataRepository.ts]: Fixed in f07a54428a. Host-scoped describe errors now record their AppHost owner, survive selection changes, and clear only when that owner recovers or leaves the desired set. CLI-wide errors remain ownerless and can still be cleared by any successful peer. The regression covers a selected-host failure with two peers, peer data not clearing the error, and owner removal clearing it.

**Comment by adamint** [comment]
I’m going to update the five stop/cancellation tests so they first drive `ps --follow` to report a matching host and then assert against the spawned `describe --follow` child; right now they only cover the ps process. I’ll also fix the two indentation slips in `AppHostDataRepository.ts`, then rerun the extension tests and real Extension Host path.

File: extension/src/test/appHostDataRepository.test.ts [hunk 10]
```diff
@@ -3098,55 +3610,107 @@ suite('AppHostDataRepository', () => {
...
-        repository.dispose();
+            assert.strictEqual(repository.workspaceResources.length, 0);
+        } finally {
+            repository.dispose();
+            workspaceFoldersStub.restore();
+        }
     });
 
-    test('hiding workspace panel before cli path resolves prevents describe watch from starting', async () => {
+    test('hiding workspace panel before cli path resolves prevents describe from starting', async () => {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/appHostDataRepository.test.ts:3846]: Fixed in d429d70205. The test now starts `ps --follow`, reports the matching running host, defers the describe CLI path, hides the panel, and verifies the pending describe never spawns.

File: extension/src/test/appHostDataRepository.test.ts [hunk 11]
```diff
@@ -3680,7 +4305,7 @@ suite('AppHostDataRepository', () => {
         }
     });
 
-    test('late close from stopped describe watch does not orphan replacement watch', async () => {
+    test('late close from stopped describe does not orphan replacement', async () => {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/appHostDataRepository.test.ts:4473]: Fixed in d429d70205. This now captures children by command, drives `ps --follow` to start a describe, replaces that describe, and fires the stale describe callbacks before checking both describe processes are stopped.

File: extension/src/test/appHostDataRepository.test.ts [hunk 12]
```diff
@@ -3708,7 +4333,7 @@ suite('AppHostDataRepository', () => {
         repository.dispose();
     });
 
-    test('stubborn describe watch is force killed', async () => {
+    test('stubborn describe is force killed', async () => {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/appHostDataRepository.test.ts:4514]: Fixed in d429d70205. The test now starts and captures a stubborn describe child before hiding the panel, so the SIGTERM-to-SIGKILL assertion covers `_stopDescribe`.

File: extension/src/test/appHostDataRepository.test.ts [hunk 13]
```diff
@@ -3729,7 +4354,7 @@ suite('AppHostDataRepository', () => {
         }
     });
 
-    test('Windows describe watch termination uses taskkill for the process tree', async () => {
+    test('Windows describe termination uses taskkill for the process tree', async () => {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/appHostDataRepository.test.ts:4543]: Fixed in d429d70205. The Windows test now starts and selects the describe child with PID 4242, so both `taskkill /t` calls cover `_stopDescribe` rather than the ps process.

File: extension/src/test/appHostDataRepository.test.ts [hunk 14]
```diff
@@ -3785,7 +4410,7 @@ suite('AppHostDataRepository', () => {
         }
     });
 
-    test('already-exited describe watch is not terminated again', async () => {
+    test('already-exited describe is not terminated again', async () => {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/test/appHostDataRepository.test.ts:4607]: Fixed in d429d70205. The test now marks the live describe child exited before hiding the panel and verifies `_stopDescribe` does not terminate it or attach exit listeners.

File: extension/src/views/AppHostDataRepository.ts [hunk 15]
```diff
@@ -931,167 +917,166 @@ export class AppHostDataRepository {
...
-
-    private _clearStoppedWorkspaceAppHost(): void {
-        const appHostPath = this._workspaceAppHost?.appHostPath ?? this._workspaceAppHostPath;
-        this._workspaceAppHost = undefined;
-        this._appHosts = appHostPath
-            ? this._appHosts.filter(appHost => !isMatchingAppHostPath(appHost.appHostPath, appHostPath))
-            : [];
+    private _scheduleDescribeRestart(appHostPath: string, stream: DescribeStream): void {
+        const delay = stream.restartDelay;
+        const nextDelay = Math.min(delay * 2, this._getPollingIntervalMs());
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/views/AppHostDataRepository.ts]: Good catch. I’ll clamp the next describe retry delay so a polling interval below five seconds cannot reduce the current backoff, and add coverage for the 1-second setting.

File: extension/src/editor/AppHostFilePresenceWatcher.ts [hunk 16]
```diff
@@ -1,48 +1,75 @@
...
+    private _lastReportedOpenPaths: readonly string[] = [];
     private _changeTimer: NodeJS.Timeout | undefined;
     private _updateVersion = 0;
     private _updateTask: Promise<void> = Promise.resolve();
 
     constructor(private readonly _repository: AppHostDataRepository) {
         this._disposables.push(
+            vscode.window.tabGroups.onDidChangeTabs(() => this._queueUpdate()),
             vscode.window.onDidChangeVisibleTextEditors(() => this._queueUpdate()),
+            vscode.window.onDidChangeActiveTextEditor(() => this._queueUpdate()),
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/AppHostFilePresenceWatcher.ts:41]: Agreed. I’ll turn the isolated Extension Host proof into checked-in E2E coverage for an out-of-workspace AppHost tab staying followed after it is backgrounded.

File: extension/src/views/AppHostDataRepository.ts [hunk 17]
```diff
@@ -1858,108 +1720,76 @@ export class AppHostDataRepository {
...
-            workspaceAppHost = workspaceAppHosts[0];
-            workspaceAppHostPathChanged = !isMatchingAppHostPath(workspaceAppHostPath, workspaceAppHost.appHostPath);
-            if (workspaceAppHostPathChanged) {
-                extensionLogOutputChannel.info(`Retargeting workspace AppHost describe to running AppHost ${workspaceAppHost.appHostPath}`);
-                this._stopDescribeWatch({ clearWorkspaceResources: true });
-                this._setWorkspaceAppHostPathFromCurrentCandidates(workspaceAppHost.appHostPath);
-                workspaceAppHostPath = this._workspaceAppHostPath;
-                this._setDescribeError(undefined);
-                this._describeRestartDelay = 5000;
+        if (!selectedWorkspaceAppHost && workspaceAppHosts.length === 1) {
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/views/AppHostDataRepository.ts:1735]: Good catch. I’ll restore a focused unified-stream regression that starts with multiple candidates, reports only the non-selected candidate running, and verifies both selection and the describe target retarget.

File: extension/src/editor/AppHostFilePresenceWatcher.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/AppHostFilePresenceWatcher.ts:41]: Added this in 554d3c1018. The checked-in Extension Host test opens an external single-file AppHost, starts it, backgrounds its real tab, hides the Aspire panel, and verifies the host/resources remain followed. The full extension suite passed 1,172 tests and the edge-case E2E suite passed 4/4.

File: extension/src/views/AppHostDataRepository.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/views/AppHostDataRepository.ts:1735]: Restored this in 554d3c1018. The focused test starts with two candidates, reports only the non-selected candidate running, and verifies both workspace selection and the describe target retarget. The full extension suite passed 1,172 tests.

**Review by adamint — APPROVED** [review]
The backoff issue and lifecycle coverage gaps are addressed. I do not see any remaining blocking issues.
